crates/rust-analyzer/src/main_loop/handlers.rs

```diff
@@ -696,54 +696,103 @@ pub fn handle_formatting(
-    let frange = FileRange { file_id, range };
 
     let diagnostics = world.analysis().diagnostics(file_id)?;
-    let mut res: Vec<lsp_ext::CodeAction> = Vec::new();
 
     let fixes_from_diagnostics = diagnostics
         .into_iter()
         .filter_map(|d| Some((d.range, d.fix?)))
         .filter(|(diag_range, _fix)| diag_range.intersect(range).is_some())
         .map(|(_range, fix)| fix);
-
     for fix in fixes_from_diagnostics {
         let title = fix.label;
         let edit = to_proto::snippet_workspace_edit(&world, fix.source_change)?;
-        let action =
-            lsp_ext::CodeAction { title, group: None, kind: None, edit: Some(edit), command: None };
+        let action = lsp_ext::CodeAction {
+            title,
+            id: None,
+            group: None,
+            kind: None,
+            edit: Some(edit),
+            command: None,
+        };
         res.push(action);
     }
-
     for fix in world.check_fixes.get(&file_id).into_iter().flatten() {
         let fix_range = from_proto::text_range(&line_index, fix.range);
         if fix_range.intersect(range).is_none() {
             continue;
         }
         res.push(fix.action.clone());
     }
+    Ok(())
+}
 
-    for assist in world.analysis().assists(&world.config.assist, frange)?.into_iter() {
-        res.push(to_proto::code_action(&world, assist)?.into());
+pub fn handle_code_action(
+    world: WorldSnapshot,
+    params: lsp_types::CodeActionParams,
+) -> Result<Option<Vec<lsp_ext::CodeAction>>> {
+    let _p = profile("handle_code_action");
+    // We intentionally don't support command-based actions, as those either
+    // requires custom client-code anyway, or requires server-initiated edits.
+    // Server initiated edits break causality, so we avoid those as well.
+    if !world.config.client_caps.code_action_literals {
+        return Ok(None);
+    }
+
+    let file_id = from_proto::file_id(&world, &params.text_document.uri)?;
+    let line_index = world.analysis().file_line_index(file_id)?;
+    let range = from_proto::text_range(&line_index, params.range);
+    let frange = FileRange { file_id, range };
+    let mut res: Vec<lsp_ext::CodeAction> = Vec::new();
+
+    handle_fixes(&world, &params, &mut res)?;
+
+    if world.config.client_caps.resolve_code_action {
+        for assist in world.analysis().unresolved_assists(&world.config.assist, frange)?.into_iter()
+        {
+            res.push(to_proto::unresolved_code_action(&world, assist)?);
+        }
+    } else {
+        for assist in world.analysis().resolved_assists(&world.config.assist, frange)?.into_iter() {
+            res.push(to_proto::resolved_code_action(&world, assist)?);
+        }
     }
+
     Ok(Some(res))
 }
 
+pub fn handle_resolve_code_action(
+    world: WorldSnapshot,
+    params: lsp_ext::ResolveCodeActionParams,
+) -> Result<Option<lsp_ext::SnippetWorkspaceEdit>> {
+    if !world.config.client_caps.resolve_code_action {
```

Comment:
I don't think we need this check here -- if the client sends a request, we might as well handle it, even if the client lied about their incapability :)